### PR TITLE
refactor pairings system

### PR DESCRIPTION
**Pull request by mgoren**

### CODE REVIEW

File: app/controllers/attendance_sign_in_remote_controller.rb [hunk 1]
```diff
@@ -12,8 +12,8 @@ def new
   def create
     record = AttendanceRecord.find_or_initialize_by(student: current_student, date: Time.zone.now.in_time_zone(current_student.course.office.time_zone).to_date)
     pair_ids = params[:pair_ids].uniq.reject {|p| p.empty?}
-    record.pair_ids = pair_ids
-    if record.save
+    record.pairings.delete_all
```
Comment (michaelrkn) [inline app/controllers/attendance_sign_in_remote_controller.rb:15]: Why is this in the controller?

File: app/controllers/pair_feedbacks_controller.rb [hunk 2]
```diff
@@ -22,19 +19,16 @@ def new
...
-        if @pairs_without_feedback.any?
+        if @student.pairs_without_feedback_today.any?
           redirect_to sign_out_path
         else
           sign_out_student
         end
       else
-        @pairs_without_feedback = pairs.select { |pair| !PairFeedback.where(created_at: today.all_day).find_by(student: @student, pair: pair)}
+        @pairs_without_feedback = @student.pairs_without_feedback_today
         render :new
```
Comment (michaelrkn) [inline app/controllers/pair_feedbacks_controller.rb:32]: Maybe redirect to /new instead of rendering :new. That way the user can hit refresh or bookmark the page and the page will load properly.

File: app/models/student.rb [hunk 3]
```diff
@@ -156,8 +154,23 @@ def submission_for(code_review)
     submissions.find { |submission| submission.code_review_id == code_review.id }
   end
 
+  def pairs(course=nil)
+    selected_attendance_records = course ? attendance_records.where("date between ? and ?", course.try(:start_date), course.try(:end_date)) : attendance_records
```
Comment (michaelrkn) [inline app/models/student.rb:158]: What happens if `start_date` or `end_date` is `nil`?

File: app/models/student.rb [hunk 4]
```diff
@@ -156,8 +154,23 @@ def submission_for(code_review)
...
+    (selected_attendance_records.map {|s| s.pairings.pluck(:pair_id)}.flatten).compact.sort
+  end
+
   def pairs_on_day(day)
```
Comment (michaelrkn) [inline app/models/student.rb:162]: Based on the name, I would expect this to return the students, not their IDs.

File: app/models/student.rb [hunk 5]
```diff
@@ -156,8 +154,23 @@ def submission_for(code_review)
...
+  end
+
+  def inverse_pairs_on_day(day)
+    Student.where(id: AttendanceRecord.where(date: day).joins(:pairings).where(pairings: { pair_id: id }).pluck(:student_id))
+  end
+
+  def pairs_without_feedback_today
+    today = Time.zone.now.to_date
+    all_pairs = pairs_on_day(today).or(inverse_pairs_on_day(today))
+    all_pairs.select { |pair| !PairFeedback.today.find_by(student: self, pair: pair)}
```
Comment (michaelrkn) [inline app/models/student.rb:173]: Can you change these two lines to use SQL/AR? It will be more efficient.

File: app/controllers/attendance_sign_in_remote_controller.rb (hunk 1 above)
Comment (mgoren) [reply app/controllers/attendance_sign_in_remote_controller.rb:15]: updating with dependent attributes actually adds them rather than replaces them... which is why deleting existing pairings is needed. I like the idea of moving that to AttendanceRecord#update_pairings or something like that tho.

File: app/controllers/pair_feedbacks_controller.rb (hunk 2 above)
Comment (mgoren) [reply app/controllers/pair_feedbacks_controller.rb:32]: I think the render :new is there so that it will render validation error messages, but really the only reason it'll fail is if one or more of the questions is unanswered, so I could just redirect with an alert. It will simplify the code too to use a redirect. I think I was just under the impression it was better to render the validation error messages when possible.

File: app/models/student.rb (hunk 3 above)
Comment (mgoren) [reply app/models/student.rb:158]: Courses always have a start_time and end_time (now calculated through class_times rather than hardcoded). The problem was specs. 

This is actually just because when I was adding class_times recently I didn't feel like I had the time to go through and fix all the specs that have courses for purposes other than testing class times. But actually I added in a stub for start_time and end_time to be used by tests where I'm not specifically testing those things, so I should be able to delete that now.

File: app/models/student.rb (hunk 4 above)
Comment (mgoren) [reply app/models/student.rb:162]: Agreed. I should change this to return students.

File: app/models/student.rb (hunk 5 above)
Comment (mgoren) [reply app/models/student.rb:173]: Yes likely! I've actually just begun to change over in a few places from using select to trying to use joins etc to be as efficient as possible.

File: app/controllers/attendance_sign_in_remote_controller.rb (hunk 1 above)
Comment (michaelrkn) [reply app/controllers/attendance_sign_in_remote_controller.rb:15]: You can also use a before_update callback.

File: app/controllers/pair_feedbacks_controller.rb (hunk 2 above)
Comment (michaelrkn) [reply app/controllers/pair_feedbacks_controller.rb:32]: Oh, no, you're right, I just misunderstood the code. It is generally better to render :new with the validation error messages. That way the form is already filled out (if there is one) from the instance variables' data.

File: app/models/student.rb (hunk 5 above)
Comment (michaelrkn) [reply app/models/student.rb:173]: Sweet!

Comment (mgoren) [reply app/models/student.rb:173]: I was able to replace this with an AR query, but it feels a bit wordy. I imagine there might be a cleaner way. I mean obviously I could combine all of this into one line, but that doesn't seem like it would move things in the direction of being clearer.

```
def pairs_without_feedback_today
  today = Time.zone.now.to_date
  all_pairs = pairs_on_day(Time.zone.now.to_date).or(inverse_pairs_on_day(Time.zone.now.to_date))
  existing_feedback = PairFeedback.today.where(student: self, pair: all_pairs)
  all_pairs.where.not(id: existing_feedback.pluck(:pair_id))
end
```

Comment (mgoren) [reply app/models/student.rb]: Actually scratch that. I realized that I no longer need to consider unreciprocated pairs for purposes of PairFeedback, because presumably by the end of the day any unreciprocated pairings should be fixed. And I realized I could clean this up to just:

```
  def pairs_without_feedback_today
    pairs_on_day(Time.zone.now.to_date).where.not(id: evaluations_of_pairs.today.pluck(:pair_id))
  end
```

File: app/models/student.rb [hunk 6]
```diff
@@ -99,11 +99,9 @@ def absences_overall
   end
 
   def solos(filtered_course = nil)
-    if filtered_course
-      attendance_records.where(pair_ids: [], ignore: nil).where("date between ? and ?", filtered_course.try(:start_date), filtered_course.try(:end_date)).select { |ar| !ar.date.friday? }.count
-    else
-      attendance_records.where(pair_ids: [], ignore: nil).select { |ar| !ar.date.friday? }.count
-    end
+    solo_records = attendance_records.includes(:pairings).where(pairings: {id: nil})
```
Comment (michaelrkn) [inline app/models/student.rb:102]: This line isn't used when `filtered_course` is provided, right?

File: app/models/student.rb [hunk 7]
```diff
@@ -99,11 +99,9 @@ def absences_overall
...
+    filtered_records = filtered_course ? solo_records.where("date between ? and ?", filtered_course.start_date, filtered_course.end_date) : solo_records
+    filtered_records.reject {|ar| ar.date.friday?}.count
```
Comment (michaelrkn) [inline app/models/student.rb:104]: `reject` is a Ruby array method, I believe. This will be more efficient if you write it with AR :)

File: app/models/student.rb [hunk 8]
```diff
@@ -156,8 +154,29 @@ def submission_for(code_review)
     submissions.find { |submission| submission.code_review_id == code_review.id }
   end
 
+  def pair_ids(course=nil)
+    selected_attendance_records = course ? attendance_records.where("date between ? and ?", course.start_date, course.end_date) : attendance_records
+    (selected_attendance_records.map {|s| s.pairings.pluck(:pair_id)}.flatten).compact.sort
```
Comment (michaelrkn) [inline app/models/student.rb:159]: Why not call `pluck` on the line above?

File: app/models/student.rb [hunk 9]
```diff
@@ -156,8 +154,29 @@ def submission_for(code_review)
...
+  def orphan_pairs_on_day(day) # extra pairs this student nonreciprocally claimed
+    pairs_on_day(day).where.not(id: inverse_pairs_on_day(day))
+  end
+
+  def inverse_orphan_pairs_on_day(day) # students who nonreciprocally claimed this student as a pair
+    inverse_pairs_on_day(day).where.not(id: pairs_on_day(day))
+  end
+
+  def pairs_without_feedback_today
+    pairs_on_day(Time.zone.now.to_date).where.not(id: evaluations_of_pairs.today.pluck(:pair_id))
```
Comment (michaelrkn) [inline app/models/student.rb:179]: All these methods above refer to "pairs", but they are returning "paid_ids".

File: app/views/students/_in_session.html.erb [hunk 10]
```diff
@@ -9,10 +9,16 @@
...
           <% if current_student.pairs_on_day(Time.zone.now.to_date).any? %>
-            You are signed in with <%= current_student.pairs_on_day(Time.zone.now.to_date).map {|s| s.name}.join(' & ') %>. [<%= link_to 'change pair', sign_in_path, remote: true, id: 'attendance-change-pair-link' %>]
+            <span id='signed-in-with'>You are signed in with <%= current_student.pairs_on_day(Time.zone.now.to_date).map {|s| s.name}.join(' & ') %>. [<%= link_to 'change pair', sign_in_path, remote: true, id: 'attendance-change-pair-link' %>]</p>
           <% else %>
             <strong class='text-primary'>You are signed in without a pair!</strong> Click <%= link_to 'here', sign_in_path, remote: true, id: 'attendance-change-pair-link' %> to log your pair.
           <% end %>
+          <% if current_student.inverse_orphan_pairs_on_day(Time.zone.now.to_date).any? %>
+            <p class='alert alert-danger'>
+              <strong>Additional students have marked you as a pair today: <%= current_student.inverse_orphan_pairs_on_day(Time.zone.now.to_date).order(:name).pluck(:name).join(' & ') %></strong><br>
```
Comment (michaelrkn) [inline app/views/students/_in_session.html.erb:18]: This is too much logic for a view. Call `current_student.inverse_orphan_pairs_today.pluck(:name).join(' & ')`.

**Review by michaelrkn — COMMENTED** [review]
Looking good, Mike!

File: app/models/student.rb (hunk 6 above)
Comment (mgoren) [reply app/models/student.rb:102]: It is used by the line that builds filtered_records, regardless of the presence of filtered_course. I refactored that method (student#solos) because I thought this way of writing it was easier to read, but now I'm not totally convinced.

File: app/models/student.rb (hunk 7 above)
Comment (mgoren) [reply app/models/student.rb:104]: agreed. will work on that.

File: app/models/student.rb (hunk 8 above)
Comment (mgoren) [reply app/models/student.rb:159]: Do you mean combine the whole method onto one line, or just move pluck up? It's split over two lines just for theoretical readability. I'm open to suggestions.

File: app/models/student.rb (hunk 9 above)
Comment (mgoren) [reply app/models/student.rb:179]: These methods do return students actually: 
pairs_on_day
inverse_pairs_on_day
orphan_pairs_on_day
inverse_orphan_pairs_on_day
pairs_without_feedback_on_day

File: app/views/students/_in_session.html.erb (hunk 10 above)
Comment (mgoren) [reply app/views/students/_in_session.html.erb:18]: Do you mean just make an additional method inverse_orphan_pairs_today that just calls inverse_orphan_pairs_on_day(Time.zone.now.to_date)?

File: app/models/student.rb (hunk 6 above)
Comment (michaelrkn) [reply app/models/student.rb:102]: Oh never mind I just didn't read the ternary statement correctly. I would suggest not using the ternary operator for a line that long.

File: app/models/student.rb (hunk 8 above)
Comment (michaelrkn) [reply app/models/student.rb:159]: Okay, again, I mis-read it. Your code is better than what I was thinking :) But use AR instead of Ruby - once you call `map` you become very inefficient. I think you want something like `selected_attendance_records.joins(:pairings).pluck(:pair_id)` although I'm sure that's not quite right.

File: app/models/student.rb (hunk 9 above)
Comment (michaelrkn) [reply app/models/student.rb:179]: Well gosh, I just saw the `pluck(:pair_id)` and didn't notice the double closing parenthesis. Oops sorry!

File: app/views/students/_in_session.html.erb (hunk 10 above)
Comment (michaelrkn) [reply app/views/students/_in_session.html.erb:18]: Yeah exactly. I dunno, maybe it's not necessary, just feels like a lot of Ruby for a view.

File: app/models/student.rb (hunk 6 above)
Comment (mgoren) [reply app/models/student.rb:102]: Makes sense. Yeah I'll replace with a simple if else.

File: app/models/student.rb (hunk 7 above)
Comment (mgoren) [reply app/models/student.rb:104]: I'm actually not sure if I can write this one as an AR query, because it needs to test whether Date is Friday, which is a Ruby method anyways. Actually I may have a better solution. Attendance records are no longer created on Fridays, and I should just be able to go back and delete all Friday attendance records from the past.

File: app/models/student.rb (hunk 8 above)
Comment (mgoren) [reply app/models/student.rb:159]: Yep I'll work on that. I only recently figured out a bit more how to use joins etc, so I'm happy to be transitioning stuff over to that.

File: app/views/students/_in_session.html.erb (hunk 10 above)
Comment (mgoren) [reply app/views/students/_in_session.html.erb:18]: Agreed. I keep considering just defaulting day to Time.zone.now.to_date on all those methods, so if calling @student.pairs_on_day it will auto-default to today if not passed a date. Does that make sense, or do you think it's better to write separate methods called pairs_today, etc, for clarity?

File: app/models/student.rb (hunk 7 above)
Comment (michaelrkn) [reply app/models/student.rb:104]: Sounds good. Also there are SQL methods for most things :) https://www.w3schools.com/sql/func_mysql_weekday.asp

File: app/views/students/_in_session.html.erb (hunk 10 above)
Comment (michaelrkn) [reply app/views/students/_in_session.html.erb:18]: I would lean towards more clarity.

File: app/models/student.rb (hunk 7 above)
Comment (mgoren) [reply app/models/student.rb:104]: I think I do want to just fix the issue where Friday attendance records are still sometimes being created, but it turns out that means fixing things in a few different places, so I'd like to just leave this until after I merge in this PR, and do it separately, since it's kinda a separate thing.
